anvil-python/anvil/commands/upgrades/inter_channel.py

```diff
@@ -0,0 +1,275 @@
+    CONFIG_KEY as HAPROXY_CONFIG_KEY,
+    HAPROXY_UNIT_TIMEOUT,
+)
+from anvil.commands.maas_agent import (
+    CONFIG_KEY as MAASAGENT_CONFIG_KEY,
+    MAASAGENT_UNIT_TIMEOUT,
+)
+from anvil.commands.maas_region import (
+    CONFIG_KEY as MAASREGION_CONFIG_KEY,
+    MAASREGION_UNIT_TIMEOUT,
+)
+from anvil.commands.postgresql import (
+    CONFIG_KEY as POSTGRESQL_CONFIG_KEY,
+    POSTGRESQL_UNIT_TIMEOUT,
+)
+from anvil.commands.upgrades.base import (
+    UpgradePlugins,
+)
+from anvil.jobs.manifest import Manifest
+
+LOG = logging.getLogger(__name__)
+console = Console()
+
+
+class UpgradeCharm(BaseStep, JujuStepHelper):
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        client: Client,
+        jhelper: JujuHelper,
+        manifest: Manifest,
+        model: str,
+        charms: list[str],
+        tfplan: str,
+        config: str,
+        timeout: int,
+    ):
+        super().__init__(name, description)
+        self.client = client
+        self.jhelper = jhelper
+        self.manifest = manifest
+        self.model = model
+        self.charms = charms
+        self.tfplan = tfplan
+        self.config = config
+        self.timeout = timeout
+
+    def run(self, status: Status | None = None) -> Result:
+        """Run machine charm upgrade."""
+        apps = self.get_apps_filter_by_charms(self.model, self.charms)
+        result = self.upgrade_applications(
+            apps,
+            self.charms,
+            self.model,
+            self.tfplan,
+            self.config,
+            self.timeout,
+            status,
+        )
+        return result
+
+    def upgrade_applications(
+        self,
+        apps: list[str],
+        charms: list[str],
+        model: str,
+        tfplan: str,
+        config: str,
+        timeout: int,
+        status: Status | None = None,
+    ) -> Result:
+        expected_wls = ["active", "blocked", "unknown"]
+        LOG.debug(
+            f"Upgrading applications using Terraform plan {tfplan}: {apps}"
+        )
+        try:
+            self.manifest.update_partial_tfvars_and_apply_tf(
+                self.client, charms, tfplan, config
+            )
+        except TerraformException as e:
+            LOG.exception("Error upgrading cloud")
+            return Result(ResultType.FAILED, str(e))
+
+        task = run_sync(update_status_background(self, apps, status))
+        try:
+            run_sync(
+                self.jhelper.wait_until_desired_status(
+                    model,
+                    apps,
+                    expected_wls,
+                    timeout=timeout,
+                )
+            )
+        except (JujuWaitException, TimeoutException) as e:
+            LOG.debug(str(e))
+            return Result(ResultType.FAILED, str(e))
+        finally:
+            if not task.done():
+                task.cancel()
+
+        return Result(ResultType.COMPLETED)
+
+
+class UpgradeHAProxyCharms(UpgradeCharm):
```

Comment:
Let's turn this class name and the relevant ones to singular:

```suggestion
class UpgradeHAProxyCharm(UpgradeCharm):
```